File: frontend/src/types/status.ts
import type { EntityConfig } from './domain';

export type UnitState = 'standby' | 'running' | 'limited' | 'stopped';
export const ALL_UNIT_STATE: readonly UnitState[] = ['standby', 'running', 'limited', 'stopped'];
export type DecisionState = 'draft' | 'review' | 'accepted' | 'escalated' | 'review_required';
export const ALL_DECISION_STATE: readonly DecisionState[] = ['draft', 'review', 'accepted', 'escalated', 'review_required'];

export const ENTITY_CONFIGS: readonly EntityConfig[] = [
  { key: 'captureUnit', path: 'units', label: '捕集装置', statuses: ['standby', 'running', 'limited', 'stopped'] as const },
  { key: 'permitRule', path: 'rules', label: '许可规则', statuses: ['draft', 'active', 'superseded', 'retired'] as const },
  { key: 'emissionSample', path: 'samples', label: '排放样本', statuses: ['collected', 'testing', 'verified', 'invalid'] as const },
  { key: 'complianceDecision', path: 'decisions', label: '合规决定', statuses: ['draft', 'review', 'accepted', 'escalated', 'review_required'] as const }
];

// Human-facing Chinese labels for machine statuses. Falls back to the raw
// status so unknown values stay visible instead of rendering blank.
const STATUS_LABELS: Record<string, string> = {
  standby: '待机', running: '运行中', limited: '限产', stopped: '停机',
  draft: '草稿', review: '复核中', accepted: '已接受', escalated: '已升级',
  review_required: '回退复核', collected: '已采样', testing: '检测中',
  verified: '已验证', invalid: '已作废', active: '生效中', superseded: '已替代', retired: '已停用',
};

export function statusLabel(status: string): string {
  return STATUS_LABELS[status] || status.replaceAll('_', ' ');
}

File: backend/internal/constants/status.go
package constants

// Shared status values are mirrored in frontend/src/types/status.ts. Keeping
// the lists explicit makes state-machine drift visible during code review.

type UnitState string

const (
	UnitStateStandby UnitState = "standby"
	UnitStateRunning UnitState = "running"
	UnitStateLimited UnitState = "limited"
	UnitStateStopped UnitState = "stopped"
)

var AllUnitState = []string{"standby", "running", "limited", "stopped"}

type DecisionState string

const (
	DecisionStateDraft          DecisionState = "draft"
	DecisionStateReview         DecisionState = "review"
	DecisionStateAccepted       DecisionState = "accepted"
	DecisionStateEscalated      DecisionState = "escalated"
	DecisionStateReviewRequired DecisionState = "review_required"
)

var AllDecisionState = []string{"draft", "review", "accepted", "escalated", "review_required"}

var CaptureUnitTransitions = map[string]map[string]bool{
	"standby": {"running": true, "limited": true},
	"running": {"limited": true, "stopped": true, "standby": true},
	"limited": {"stopped": true, "running": true},
	"stopped": {"limited": true},
}

var PermitRuleTransitions = map[string]map[string]bool{
	"draft":      {"active": true, "superseded": true},
	"active":     {"superseded": true, "retired": true, "draft": true},
	"superseded": {"retired": true, "active": true},
	"retired":    {"superseded": true},
}

var EmissionSampleTransitions = map[string]map[string]bool{
	"collected": {"testing": true, "verified": true},
	"testing":   {"verified": true, "invalid": true, "collected": true},
	"verified":  {"invalid": true, "testing": true},
	"invalid":   {"verified": true},
}

// ComplianceDecisionTransitions covers the regular review workflow. The
// review_required state is entered only by the sample-void rollback flow and
// left only by a substitute-sample final review, so it deliberately has no
// edges in this graph: generic transition requests to or from it are rejected.
var ComplianceDecisionTransitions = map[string]map[string]bool{
	"draft":     {"review": true},
	"review":    {"accepted": true, "escalated": true, "draft": true},
	"accepted":  {"escalated": true, "review": true},
	"escalated": {"accepted": true},
}

// FinalDecisionStates are the decisions whose conclusion is actively in force
// and therefore must be rolled back when their emission sample is voided.
var FinalDecisionStates = []string{"accepted", "escalated"}

func CanTransition(graph map[string]map[string]bool, from, to string) bool {
	targets, exists := graph[from]
	return exists && targets[to]
}
